lib/vdb/vdb.cpp

```diff
@@ -0,0 +1,129 @@
+#include "pms-utils/vdb/vdb.hpp"
+
+#include "pms-utils/atom/atom.hpp"
+#include "pms-utils/depend/depend.hpp"
+#include "pms-utils/depend/depend_parser.hpp"
+#include "pms-utils/ebuild/ebuild.hpp"
+#include "pms-utils/misc/try_parse.hpp"
+#include "pms-utils/vdb/vdb_parsers.hpp"
+#include "util/readfile.hpp"
+
+#include <boost/parser/parser.hpp>
+#include <cstring>
+#include <expected>
+#include <filesystem>
+#include <format>
+#include <stdexcept>
+#include <string_view>
+
+using pms_utils::misc::try_parse;
+using pms_utils::misc::_internal::try_readfile;
+using pms_utils::vdb::_internal::Error;
+using pms_utils::vdb::_internal::IoError;
+using pms_utils::vdb::_internal::ParseError;
+
+namespace [[gnu::visibility("default")]] pms_utils {
+namespace vdb {
+
+namespace {
+
+template <typename T, typename E> T unwrap(std::expected<T, E> result, std::string_view message) {
+    if (!result) {
+        throw std::runtime_error(std::format("{}: {}", message, result.error()));
+    }
+
+    return result.value();
+}
+
+} // namespace
+
+Entry::Entry(const std::filesystem::path &path) {
+    {
+        for (const auto kind :
+             {std::string_view("DEPEND"), std::string_view("BDEPEND"), std::string_view("RDEPEND")}) {
+            auto depend_path = path / kind;
+
+            if (std::filesystem::exists(depend_path)) {
+                auto depend =
+                    unwrap(try_readfile(depend_path).ok_or(Error{IoError{}}).and_then([](const auto &input) {
+                        return try_parse(input, pms_utils::parsers::depend::nodes, true)
+                            .transform_error([&input](const auto err) {
+                                return Error{ParseError{std::string(input.cbegin(), err)}};
+                            });
+                    }),
+                           std::format("failed to load {}", kind));
+
+                if (kind == "DEPEND") {
+                    depend_ = depend;
+                }
+
+                if (kind == "BDEPEND") {
+                    bdepend_ = depend;
+                }
+
+                if (kind == "RDEPEND") {
+                    rdepend_ = depend;
+                }
+            }
+        }
+    }
+
+    {
+        auto keywords_path = path / "KEYWORDS";
+
+        if (std::filesystem::exists(keywords_path)) {
+            keywords_ =
+                unwrap(try_readfile(keywords_path).ok_or(Error{IoError{}}).and_then([](const auto &input) {
+                    return try_parse(input, pms_utils::parsers::vdb::_internal::VDB_KEYWORDS, true)
+                        .transform_error([&input](const auto err) {
+                            return Error{ParseError{std::string(input.cbegin(), err)}};
+                        });
+                }),
+                       "failed to load KEYWORDS")
+                    .value_or(pms_utils::ebuild::keywords{});
```

Comment:
`.or_else( []() { return keywords{}; } );` so you don't construct the temporary unless needed